libevmasm/Assembly.cpp

```diff
@@ -1393,6 +1398,24 @@ LinkerObject const& Assembly::assembleEOF() const
 	m_tagPositionsInBytecode = std::vector<size_t>(m_usedTags, std::numeric_limits<size_t>::max());
 	std::map<size_t, uint16_t> dataSectionRef;
 
+	bool storesAuxData = false;
+	bool loadsAuxData = false;
+
+	for (auto& codeSection: m_codeSections)
+	{
+		for (auto const& item: codeSection.items)
+			if (item.type() == AuxDataStore)
+				storesAuxData = true;
+			else if (item.type() == AuxDataLoadN)
+				loadsAuxData = true;
+		if (storesAuxData || loadsAuxData)
+			assertThrow(
+				storesAuxData != loadsAuxData,
+				AssemblyException,
+				"Cannot store and load auxilairy data in the same assembly subroutine."
+			);
+	}
```

Comment:
You asked about testing this check and there's really no good way to do it directly, because any checks here are just the last line of defense. They should be detected during analysis instead, where we have a proper error reporting mechanism.

The thing is, we don't have analysis stage for assembly, since we don't support it as input. We can get here only via Yul or through assembly import. So this check should really be at Yul level (probably in `AsmAnalysis.cpp`) and later be also added as validation in the import. Here it should just be an assert (and I'm not sure this is really the right spot for this assert, since we have to scan the whole bytecode to check it, maybe we can detect it when processing `AuxDataStore` and/or `AuxDataLoadN` instead?).

Reply:
BTW, was the problem with testing it that `AssemblyException` crashes the compiler instead of showing up as an error in syntax tests? I'm not sure we're actually catching `AssemblyException` up the stack, but we definitely should.

TBH, most of the errors in `Assembly.cpp` should instead be assertions since they are things that should be caught by analysis. There are probably only a few that can't and have a legit reason to have validations here. And then they should be presented to to user as as errors, not crashes.